### PR TITLE
[do not merge] Initial design for generic PageEnumerable and PageEnumerator

### PR DESCRIPTION
**Pull request by cormacpayne**
## Description

Refactored `PageEnumerable` and `PageEnumerator` so that they can accept generic `List` and `ListNext` methods from clients.

Implemented `SupportsPaging` for `Get-AzureRmADApplication` to ensure that the changes to `PageEnumerable` and `PageEnumerator` could be accurately mirrored using `Skip` and `First` in the cmdlet. This change fixes issue https://github.com/Azure/azure-powershell/issues/3928

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by cormacpayne** [comment]
@DeepakRajendranMsft @EvgenyAgafonchikov please have a look

**Comment by cormacpayne** [comment]
@markcowl I need to update the change log and markdown help to reflect the changes made to `Get-AzureRmADApplication`

File: src/ResourceManager/Resources/Commands.Resources/Models.ActiveDirectory/ActiveDirectoryClient.cs [hunk 1]
```diff
@@ -705,9 +706,13 @@ public PSADApplication GetApplication(string applicationObjectId)
             return GraphClient.Applications.Get(applicationObjectId.ToString()).ToPSADApplication();
         }
 
-        public IEnumerable<PSADApplication> GetApplicationWithFilters(Rest.Azure.OData.ODataQuery<Application> odataQueryFilter)
+        public IEnumerable<PSADApplication> GetApplicationWithFilters(Rest.Azure.OData.ODataQuery<Application> odataQueryFilter, ulong first = ulong.MaxValue, ulong skip = 0)
         {
-            return GraphClient.Applications.List(odataQueryFilter).Select(a => a.ToPSADApplication());
+            return new PageEnumerable<Application>(
+                delegate()
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/Models.ActiveDirectory/ActiveDirectoryClient.cs]: Use ``` () => GraphClient.Applications.List(odataQueryFilter)```

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerator.cs [hunk 2]
```diff
@@ -17,23 +17,36 @@
 using System.Collections.Generic;
 using System.Collections;
 
-namespace Microsoft.Azure.Internal.Subscriptions.Models
+namespace Microsoft.Azure.Commands.ResourceManager.Common.Paging
 {
-    public class PageEnumerator : IEnumerator<Subscription>
+    public class PageEnumerator<T> : IEnumerator<T>
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerator.cs]: Depending on what we hear from powershell core, you will need to place a constraint here, like ```where T : IPSPagingOutput```

**Pull request closed without merging**